test/unit/puppetlabs/jruby_utils/lockable_ref_pool_test.clj

```diff
@@ -0,0 +1,638 @@
+(ns puppetlabs.jruby_utils.lockable-ref-pool-test
+  (:require [clojure.test :refer :all]
+            [puppetlabs.services.jruby-pool-manager.jruby-testutils :as jruby-testutils])
+  (:import (com.puppetlabs.jruby_utils.pool ReferencePool)
+           (java.util.concurrent TimeUnit ExecutionException TimeoutException)))
+
+(defn timed-deref
+  [ref]
+  (deref ref 10000 :timed-out))
+
+(defn create-empty-pool
+  ([] (create-empty-pool 10))
+  ([maxBorrows]
+   (ReferencePool. maxBorrows)))
+
+(defn create-populated-pool
+  ([] create-populated-pool 10)
+  ([size]
+   (let [pool (create-empty-pool size)]
+     (.register pool (str "foo"))
+     pool)))
+
+(defn borrow-n-instances
+  [pool n]
+  (doall (for [_ (range n)]
+           (.borrowItem pool))))
+
+(defn return-instances
+  [pool instances]
+  (doseq [instance instances]
+    (.releaseItem pool instance)))
+
+(deftest pool-register-above-maximum-throws-exception-test
+  (testing "attempt to register new instance with pool at max capacity fails"
+    (let [pool (create-empty-pool)]
+      (.register pool "foo ok")
+      (is (thrown? IllegalStateException
+                   (.register pool "foo bar"))))))
+
+(deftest pool-lock-is-blocking-until-borrows-returned-test
+  (let [pool (create-populated-pool 3)
+        instances (borrow-n-instances pool 3)
+        future-started? (promise)
+        lock-acquired? (promise)
+        unlock? (promise)]
+    (is (= 3 (count instances)))
+    (is (not (.isLocked pool)))
+    (let [lock-thread (future (deliver future-started? true)
+                              (.lock pool)
+                              (deliver lock-acquired? true)
+                              @unlock?
+                              (.unlock pool)
+                              true)]
+      @future-started?
+      (testing "pool.lock() blocks until all instances are returned to the pool"
+        (is (not (realized? lock-acquired?)))
```

Comment:
Seems like this could race with the `lock-thread`..

Reply:
Ah, I know why it doesn't. In the let at the top, we borrow 3 of the instances. The lock can't proceed while the instances are borrowed.

Reply:
so then what happens when `deliver lock-acquired? true` is called and the instances aren't released yet?

Reply:
`(.lock pool)` will block until all of the instances are released, then proceed to acquire the lock. So we won't deliver the promise until after the lock has succeeded.

Reply:
ahhh right right, gotcha

Reply:
oh wow that is literally what it says it's testing 🤦‍♀

Reply:
Still good to understand where all the actual blocking and proceeding is taking place! It's definitely not obvious if you're not super familiar with the pool's methods.